tools/online_editor/src/lsp.ts

```diff
@@ -0,0 +1,442 @@
+            }
+        };
+    }
+
+    set canvas_id(id: string | null) {
+        this.#canvas_id = id;
+    }
+
+    get canvas_id() {
+        return this.#canvas_id;
+    }
+
+    private async render(
+        style: string,
+        source: string,
+        base_url: string,
+        load_callback: (_url: string) => Promise<string>,
+    ): Promise<monaco.editor.IMarkerData[]> {
+        if (this.#canvas_id == null) {
+            return Promise.resolve([]);
+        }
+
+        const { component, diagnostics, error_string } =
+            await slint_preview.compile_from_string_with_style(
+                source,
+                base_url,
+                style,
+                load_callback,
+            );
+
+        this.#client_port.postMessage({
+            type: "ErrorReport",
+            data: error_string,
+        });
+
+        const markers = diagnostics.map(function (x) {
+            return {
+                severity: 3 - x.level,
+                message: x.message,
+                source: x.fileName,
+                startLineNumber: x.lineNumber,
+                startColumn: x.columnNumber,
+                endLineNumber: x.lineNumber,
+                endColumn: -1,
+            };
+        });
+
+        if (component != null) {
+            // It's not enough for the canvas element to exist, in order to extract a webgl rendering
+            // context, the element needs to be attached to the window's dom.
+            if (this.#instance == null) {
+                this.#instance = component.create(this.canvas_id!); // eslint-disable-line
+                this.#instance.show();
+                try {
+                    if (!is_event_loop_running) {
+                        slint_preview.run_event_loop();
+                        // this will trigger a JS exception, so this line will never be reached!
+                    }
+                } catch (e) {
+                    // The winit event loop, when targeting wasm, throws a JavaScript exception to break out of
+                    // Rust without running any destructors. Don't rethrow the exception but swallow it, as
+                    // this is no error and we truly want to resolve the promise of this function by returning
+                    // the model markers.
+                    is_event_loop_running = true; // Assume the winit caused the exception and that the event loop is up now
+                }
+            } else {
+                this.#instance = component.create_with_existing_window(
+                    this.#instance,
+                );
+            }
+        }
+
+        return Promise.resolve(markers);
+    }
+
+    private highlight(file_path: string, offset: number) {
+        this.#to_highlight = { file: file_path, offset: offset };
+        this.#instance?.highlight(file_path, offset);
+    }
+}
+
+// TODO: Remove this again and hide this behind the LSP.
+export class Previewer {
```

Comment:
I wonder why the Previewer is in this file.  Should there not be a preview.ts or something like that instead?

I'm not sure what "hide behind the LSP" means